### DISCUSSION TITLE
Message extraction

### DISCUSSION THREAD

**Discussion by martinsnajdr** [amannn/next-intl#456] [Q&A]
Hello,

what's the recommended approach to extract messages from the source code and generate the JSON automatically?

Thank you.

**Comment by godfrednanaowusu** [comment]
why is this not a priority, large projects need this

**Reply by amannn** [reply]
Can you include some background on what "message extraction" means to you and why it would be beneficial for you?

Note that you can extract messages in your code base by using [the VSCode integration via i18n Ally](https://next-intl-docs.vercel.app/docs/workflows/vscode-integration). This also allows to edit messages right from the call site.

**Reply by 413n** [reply]
Would be really nice to have a CLI command that extracts *all* the translations in the whole repository (depending on a config file that says where it should look for translations) and puts them in the default language json file. Also it tells you how many translations are not yet translated in all not-default languages (so the developer remembers to do that ahah)

Or maybe there is another way to do it without a cli?

**Reply by t18n** [reply]
@amannn it is [this feature](https://www.i18next.com/how-to/extracting-translations#id-2-using-an-extraction-tool) that we are talking about. I don't think relying on a code editor plugin is a long-term solution, since many of us using different IDEs (Vim, Zed, Helix, WebStorm, etc.) that doesn't have access to that plugin.

**Reply by amannn** [reply]
Right, I agree there's some limitation—i18nAlly is only supported in VSCode and JetBrains. If you're using other editors, extraction currently needs to be done manually.

I'm happy to use this discussion to see if there's more interest from the community in this. I guess this could be implemented via a Babel / SWC plugin, but there might be some edge cases which are somewhat tough to support. For now, I don't have priorities to pursue this. I guess this could also be implemented in userland if someone has a strong need.

**Reply by ReesMorris** [reply]
We ended up implementing a (rather hacky) solution to mimic the `react-intl` extraction method after one of our `en.json` files hit 3,000+ lines. It was becoming practically impossible to maintain.

Essentially we wrapped `useTranslations` in a custom `useIntl` hook, exposing it as `formatMessage` which accepts a `defaultMessage` and `description`, e.g:

```
formatMessage({
  defaultMessage: '{first} et al.',
  description: 'Multiple authors'
}, { first })
```

These two values get hashed to generate an `id` which is what `next-intl` is told the key is for the JSON entry. To populate the JSON file we run `i18n` which is a custom script to extract all strings out of the `formatMessage` declaration. We pretty much did the same for `getTranslations` as well.

[Here's a Gist](https://gist.github.com/ReesMorris/e913ab6ca4c1ff18e86453fea26a1219) of the above, but use at your own peril as it's definitely not optimised (e.g. hashing every single string at runtime, and having to run `i18n` to update the JSON files), but we've had no complaints about it. Would be great to see something similar baked in at some point down the line, though.

**Reply by vasanthid** [reply]
@amannn  I'm also interested in a CLI command feature that can scan all files, extract strings, and place them in the correct language files, similar to what's offered by [i18next-parser](https://github.com/i18next/i18next-parser).

Our project, a Next.js 14 app router without i18n routing, is quite extensive and might expand further as we work with distributed teams. We're currently assessing i18n libraries and were leaning towards next-intl, but the absence of an extraction tool might lead us to consider other options.

 I have a couple of questions: 
 a) Could you suggest any tools or methods for automating the extraction process in large projects, aside from VS Code plugins? 
 b) Is there a possibility that this feature could be given priority in the future?

**Reply by trangcongthanh** [reply]
@amannn At my work, we have dedicated translation teams for each project. Some of them are 3rd party services, so we can't give them access to our codebase to translate our app. It's not just a limitation for developers because of the editor plugin; it also blocks our ability to outsource the translation to other teams 😿.

**Reply by amannn** [reply]
@trangcongthanh The JSON files that `next-intl` typically operates with can be uploaded to any translation management system, making it possible for other project members like translators to collaborate (see for example: [Localization management with Crowdin](https://next-intl-docs.vercel.app/docs/workflows/localization-management)).

**Comment by michalpulpan** [comment]
👋 
Since this was a required feature for our project and we really like usability of the `next-intl`, I decided to write a [simple script](https://gist.github.com/michalpulpan/40923fc940725615b6218fdb91f397a7) for this functionality this morning.
Feel free to check it out, I tested it on our project and it seems to work fine. However, for the first run, make sure to backup your `messages` folder (even though script creates timestamped backups).

The structure of messages in our project looks like this: `messages/{LOCALE}/{NAMESPACE}.json`:

```bash
.
└── messages/
    ├── cs/
    │   ├── blog.json
    │   └── common.json
    ├── en/
    │   ├── blog.json
    │   └── common.json
    └── de/
        ├── blog.json
        └── common.json
```

So feel free to adjust it to your use-case.

**Reply by vasanthid** [reply]
@michalpulpan That's great. Thank you so much. I'll test it and let you know.

**Comment by amannn** [comment]
@felix-quotez just shared his message extraction strategy in https://github.com/amannn/next-intl/pull/1700#issuecomment-2637504341:

```tsx
import { defineMessages } from '@tw/intl/define-messages'
import { useTranslations } from '@tw/intl/use-translations'

defineMessages({
 'a.b': {
    id: 'a.b',
    description: 'foo',
    defaultMessage: 'bar',
  },
})

export function Component() {
  const t = useTranslations('a')
  return t('b')
}
```

`defineMessages` uses `@formatjs/cli` to extract messages and puts it into a format for `next-intl`.

**Reply by priyajeet** [reply]
Unlike react-intl, where build time plugin can remove, won't in this case the bundle now include english strings by default? And even worse, include descriptions and ids.

**Reply by amannn** [reply]
It might, yes. I can recommend subscribing to https://github.com/amannn/next-intl/issues/1904, I'm looking to find a built-in solution for `next-intl` that works well. Next step will be publishing an RFC, would be happy to have your feedback there! ([teaser](https://x.com/jamannnnnn/status/1960262891380031764))

**Reply by morriq** [reply]
defineMessages requires `use client` in component

**Comment by amannn** [comment]
An argument that @remcohaszing has just [brought up](https://github.com/amannn/next-intl/pull/1814#issuecomment-2758581132) for message extraction: It helps to purge unused messages automatically.

Another thought I had recently is that maybe defining messages inline could be helpful in regard to AI generation of code, similar to how TailwindCSS works really well with these tools. But AI tools are also improving and multi file editing is already a possibility, so not sure if this will be a benefit in the long run.

One aspect surely is though that defining messages inline is a bit more friendly for quick prototyping.

I haven't made my mind up yet on this subject, for the time being I'm happy to discuss experiences that people have made and to explore the pro's and con's of this approach—also potentially community-based solutions.

**Reply by remcohaszing** [reply]
When I was using `react-intl`, I really liked that I could define messages inline. Something I liked less is the use of a hash as an id. It means the message source is not clear from the translation file. Also changing a message, changes the hash. That means that when sorting, a message change causes it to be moved to somewhere else entirely.

I also really liked [`extract-react-intl-messages`](https://github.com/akameco/extract-react-intl-messages) / [`babel-plugin-react-intl-auto`](https://github.com/akameco/babel-plugin-react-intl-auto). But when I switched from Babel to TypeScript, it was really inconvenient to depend on a Babel plugin.

**Comment by amannn** [comment]
Since this topic keeps coming up from time to time, there's an issue for this topic now here: https://github.com/amannn/next-intl/issues/1904. Please leave an upvote there if this is a feature you'd like to see!

**Comment by amannn** [comment]
I'm happy to share a first draft of an RFC I've written up for first-class message extraction support in `next-intl`:

→ [Rendered version](https://github.com/amannn/next-intl/blob/feat/message-extraction/rfcs/001-message-extraction.md)
→ [Discussion](https://github.com/amannn/next-intl/discussions/2036)

I think this API can greatly improve the UX of working with translations, but there are quite a few aspects to get right here. If you're interested in this feature, I'd be really happy to hear your thoughts on the proposal!

(Please leave your feedback in the linked discussion instead of here)
